### PR TITLE
Allow deploy native contracts in any block (only once)

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/2055

### CODE REVIEW

File: src/neo/SmartContract/ApplicationEngine.cs [hunk 1]
```diff
@@ -73,7 +73,10 @@ protected internal void AddGas(long gas)
         {
             GasConsumed = checked(GasConsumed + gas);
             if (GasConsumed > gas_amount)
+            {
+                if (Snapshot?.PersistingBlock?.Index == 0) return;
```
Comment (shargon) [inline src/neo/SmartContract/ApplicationEngine.cs:77]: If we add a new syscall for each deploy we can remove this line.

**Comment by erikzhang** [comment]
I prefer to use `System.Contract.Create`.

And how do `onPersist()` and `postPersist()` work?

**Comment by Qiao-Jin** [comment]
I think logic of onPersistScript and postPersistScript should also be changed.
Besides, do we need some checks, i.e. CheckCommittee for `Neo_Native_Deploy`?

**Comment by shargon** [comment]
> I prefer to use System.Contract.Create.

Remove `Neo_Native_Deploy` and generate a native `NEFFile`?
On persist logic can be changed, but it's easy to do a conditional. `if (height > X) onPersist=...`

**Comment by erikzhang** [comment]
> Remove Neo_Native_Deploy and generate a native NEFFile?

Yes. But let's wait for #2044.

**Comment by Tommo-L** [comment]
Ping

We need to merge this pr in preview4.

**Comment by shargon** [comment]
@Tommo-L We need https://github.com/neo-project/neo/pull/2044 I will move the code to `onPersist`

**Comment by shargon** [comment]
Ready to review again

**Comment by Tommo-L** [comment]
> > Remove Neo_Native_Deploy and generate a native NEFFile?
> 
> Yes. But let's wait for #2044.

@shargon miss nef file?

**Comment by Tommo-L** [comment]
> Remove Neo_Native_Deploy and generate a native NEFFile?

If so, we should add `Update/Destory` methods in native contract.

**Comment by shargon** [comment]
`OnPersist` has no Tx, and the hash it's related to the sender, so we can't use `Contract.Create` like this...

**Comment by Tommo-L** [comment]
> `OnPersist` has no Tx, and the hash it's related to the sender, so we can't use `Contract.Create` like this...

I think we don't need to deploy it in `OnPersis`. If we use `Contract.Create`, just execute the `tx.script` is enough.

**Comment by shargon** [comment]
@Tommo-L Then we need GAS in the deploy transaction, and the sender has no gas....

**Comment by Tommo-L** [comment]
> @Tommo-L Then we need GAS in the deploy transaction, and the sender has no gas....

No need to charge for the txs in genesis block.

**Comment by erikzhang** [comment]
> If so, we should add Update/Destroy methods in native contract.

Why do we need `Update/Destroy`?

**Comment by Tommo-L** [comment]
> > If so, we should add Update/Destroy methods in native contract.
> 
> Why do we need `Update/Destroy`?

Just like a normal contract upgrade, we don't have to worry about the impact of the state, due to the native contract upgrade.

**Comment by erikzhang** [comment]
But the code is not in the `script`. It is in the native code.

File: src/neo/SmartContract/ApplicationEngine.Native.cs [hunk 2]
```diff
@@ -6,24 +6,24 @@ namespace Neo.SmartContract
 {
     partial class ApplicationEngine
     {
-        public static readonly InteropDescriptor Neo_Native_Deploy = Register("Neo.Native.Deploy", nameof(DeployNativeContracts), 0, CallFlags.AllowModifyStates, false);
+        public static readonly InteropDescriptor Neo_Native_Deploy = Register("Neo.Native.Deploy", nameof(DeployNativeContract), 0, CallFlags.AllowModifyStates, false);
```
Comment (Tommo-L) [inline src/neo/SmartContract/ApplicationEngine.Native.cs:9]: Do we need `Neo.Native.Update` ? How do we update native contract?

Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Native.cs:9]: I think that we should do a condition inside the method, and change the method hash in order to include the arguments, in this way we can have method overloads.

**Comment by roman-khimov** [comment]
I think this can be simplified by using the following model:
 * every native contract has an `onPersist` method
 * `onPersistScript` invokes all of them (iterating over `NativeContract.Contracts`, in their natural ID-based order)
 * in `onPersist` handler native contract can check for `PersistingBlock.Index == 0` (or any other block number where this contract should be activated) and deploy itself if needed

This:
 * allows to avoid `if block.Index ...` logic in `Persist` (think of new contracts added and new branches there), moving it to specific contracts
 * allows for contract to actually destroy itself if needed at certain height
 * follows #1913
 * allows for #2000 (contract-managing contract will be the first one to execute in genesis `onPersist`, so it can deploy itself using internal methods and make it possible for other contracts to use its public methods afterwards)

**Comment by erikzhang** [comment]
Please check #2119

**Pull request closed without merging**